data/lua/wml-tags.lua

```diff
@@ -462,6 +462,7 @@ end
 
 function wml_actions.terrain(cfg)
 	local terrain = cfg.terrain or wml.error("[terrain] missing required terrain= attribute")
+	if #wesnoth.terrain_types[terrain].id == 0 then wml.error("[terrain] invalid terrain="..terrain) end
```

Comment:
This looks more like a workaround. It appears that unlike `wesnoth.unit_types` `wesnoth.terrain_types` returns something not `nil` regardless if the type exists or not. IMO that should be fixed and then this test can be done properly (unless there already is another way).

Reply:
I agree that it makes sense for `wesnoth.terrain_types` to return `nil` for an unrecognized terrain code, but it should be recognized that (if I recall correctly) it's supposed to return a synthesized terrain type if passed a combined base and overlay code – so for example `wesnoth.terrain_types["Gg^Fp"]` should not be `nil` even though `Gg^Fp` not defined in the WML.

Reply:
Btw, `wesnoth.terrain_types["Gg^Fp"].id` is `grassland^pine_forest`, that seems weird. Not sure that is useful anywhere.

Reply:
that makes me think if `wesnoth.terrain_types["www^www"].id` would be `^` and need this validation to check for exactly '' and '^'

Reply:
Presumably because it's a synthesized terrain type. I don't know if the synthesized ID is useful but I guess it needs _some_ ID?

Reply:
With #8801 merged this can be done properly now.

Reply:
after 1.19.1 then, until then it is not possible to test

Reply:
You are on windows, right? You could download the installer from that PR. https://github.com/wesnoth/wesnoth/actions/runs/8796358421/artifacts/1438471252

Reply:
Required __COMPAT_LAYER=RUNASINVOKER to run it, but seems to work.

Reply:
```
$ not not wesnoth.terrain_types["Vu"]
false
$ not not wesnoth.terrain_types["^Vu"]
true
```
Should both be valid for layer=overlay purpose?

Reply:
Is it possible for `Vv` and `^Vv` to be two different terrains?

Reply:
Yes.